relay-spans/src/name.rs

```diff
@@ -0,0 +1,116 @@
+use relay_conventions::name_for_op_and_attributes;
+use relay_event_schema::protocol::Span;
+use relay_protocol::{Getter, GetterIter, Val};
+
+/// Constructs a name attribute for a span, following the rules defined in sentry-conventions.
+pub fn name_for_span(span: &Span) -> Option<String> {
+    let op = span.op.value()?;
+
+    let Some(data) = span.data.value() else {
+        return Some(name_for_op_and_attributes(op, &EmptyGetter {}));
+    };
+
+    Some(name_for_op_and_attributes(
+        op,
+        // SpanData's Getter impl treats dots in attribute names as object traversals.
+        // They have to be escaped in order for an attribute name with dots to be treated as a root
+        // attribute.
+        &EscapedGetter { getter: data },
+    ))
+}
+
+struct EmptyGetter {}
+
+impl Getter for EmptyGetter {
+    fn get_value(&self, _path: &str) -> Option<Val<'_>> {
+        None
+    }
+}
+
+struct EscapedGetter<'a> {
+    getter: &'a dyn Getter,
+}
```

Comment:
nit: Doesn't need dynamic dispatch.

```suggestion
struct EscapedGetter<'a, T: Getter>(&'a T);
```

Reply:
Done!